### DISCUSSION TITLE
(moved from ORM main branch dicussion) Sending/Receiving Data via Bluetooth with Mobile Apps (Kinomap) #161

### DISCUSSION THREAD

**Discussion by Gordon-Shumway2** [JaapvanEkris/openrowingmonitor#39] [General]
ORM v1beta_updates branch was installed on a Pi 4b.

Using  Kinomap (and other unknown applications) on an apple iPhone 15 on iOS 17.3.1 it is not possible to transmit data between ORM and Kinomap. ORM is detected via bluetooth, connection between ORM and Kinomap is established but no data is exchanged. 

The output of "dmesg|egrep -i bluetooth" shows some lines at (presumably) the time when data is supposed to be exchanged:

> [   15.648222] Bluetooth: Core ver 2.22
[   15.648351] NET: Registered PF_BLUETOOTH protocol family
[   15.648359] Bluetooth: HCI device and connection manager initialized
[   15.648379] Bluetooth: HCI socket layer initialized
[   15.648386] Bluetooth: L2CAP socket layer initialized
[   15.648403] Bluetooth: SCO socket layer initialized
[   15.666168] Bluetooth: HCI UART driver ver 2.3
[   15.666189] Bluetooth: HCI UART protocol H4 registered
[   15.666275] Bluetooth: HCI UART protocol Three-wire (H5) registered
[   15.666490] Bluetooth: HCI UART protocol Broadcom registered
[   16.062939] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[   16.062971] Bluetooth: BNEP filters: protocol multicast
[   16.062988] Bluetooth: BNEP socket layer initialized
[   16.075683] Bluetooth: MGMT ver [1.22]
[   16.324903] Bluetooth: RFCOMM TTY layer initialized
[   16.324942] Bluetooth: RFCOMM socket layer initialized
[   16.324967] Bluetooth: RFCOMM ver 1.11
[  240.955169] Bluetooth: hci0: ACL packet for unknown connection handle 13
[  241.139568] Bluetooth: hci0: ACL packet for unknown connection handle 13
[  241.149628] Bluetooth: Unexpected continuation frame (len 0)
[  241.513356] Bluetooth: Unexpected continuation frame (len 0)
[  241.538375] Bluetooth: Unexpected continuation frame (len 0)

This log was taken from a fresh OS installation with the v1beta_updates branch at some point in early February. At the point there were issues with bluetooth libraries (see https://github.com/JaapvanEkris/openrowingmonitor/discussions/35) which broke HRM functionality.
One way to verify this is by logging in to ORM, change directory to "/opt/openrowingmonitor/node_modules/@abandonware" and running "npm list". This should show the bluetooth library versions.

Those issues were remedied by a downgrade of bluetooth libraries within the repository and an upgrade on the client side.

I would recommend to upgrade the current ORM installation (which actually will downgrade some libraries) by running 

           /opt/openrowing/bin/updateopenrowingmonitor.sh 

Just out of curiosity: prior to running the upgrade script, please run the verify steps mentioned above and upload the output of "npm list" from directory "/opt/openrowingmonitor/node_modules/@abandonware".

**Comment by rywell** [comment]
Finally able to get back to this. After running the update, it says I am currently running the latest version.

Below is the output of `npm list` 

> ├── @abandonware/bleno@0.5.1-4
├── @abandonware/bluetooth-hci-socket@0.5.3-7
├── @abandonware/noble@1.9.2-15
├── @babel/eslint-parser@7.23.10
├── @babel/plugin-proposal-decorators@7.23.9
├── @babel/preset-env@7.23.9
├── @rollup/plugin-babel@6.0.4
├── @rollup/plugin-commonjs@25.0.7
├── @rollup/plugin-node-resolve@15.2.3
├── @snowpack/plugin-babel@2.1.7
├── @web/rollup-plugin-html@2.1.2
├── axios@1.6.7
├── chart.js@4.4.1
├── chartjs-plugin-datalabels@2.2.0
├── eslint-config-standard@17.1.0
├── eslint-plugin-import@2.29.1
├── eslint-plugin-lit@1.11.0
├── eslint-plugin-n@16.6.2
├── eslint-plugin-promise@6.1.1
├── eslint-plugin-wc@2.0.4
├── eslint@8.56.0
├── finalhandler@1.2.0
├── form-data@4.0.0
├── http2-proxy@5.0.53
├── incyclist-ant-plus@0.3.1
├── lit@2.8.0
├── loglevel@1.9.1
├── markdownlint-cli2@0.12.1
├── node-gyp@10.0.1
├── nodemon@3.0.3
├── nosleep.js@0.12.0
├── npm-run-all@4.1.5
├── pigpio@3.3.1
├── rollup-plugin-summary@1.4.3
├── rollup-plugin-terser@7.0.2
├── rollup@2.79.1
├── serve-static@1.15.0
├── simple-git-hooks@2.9.0
├── snowpack@3.8.8
├── tar@6.2.0
├── uvu@0.5.6
├── ws@8.16.0
└── xml2js@0.6.2

**Reply by JaapvanEkris** [reply]
This is the most recent package list, so that is OK.

**Comment by JaapvanEkris** [comment]
> I understand that it's able to connect. I recommend erf as it is able to show the data that is being broadcasted. We need to see where things go wrong. Ftms is at spec so it should work.
>
> As you said I am not able to reproduce this either which is utterly strange

Yeah, I'm using the FTMS interface with EXR on a daily basis, which works flawlessly. 

I just ran KinoMap (as I have no account I can't see how it works with a video), but when I look at the device setup, it reports its data. How does ORM report itself (what name does it use?).

Do you have anything bluetooth installed on the Raspberry Pi that might interfere?

**Reply by Gordon-Shumway2** [reply]
Judging from this Kinomap screenshot in the original discussion thread ORM identifies itself as an FTMS device of type rowing machine with the name "rowingmonitor". This looks identical to what I saw when testing on an iphone.
![kinomap](https://github.com/JaapvanEkris/openrowingmonitor/assets/78542107/da2170ce-dfcf-4d71-a3e9-55b3638618f5)

**Reply by JaapvanEkris** [reply]
yeah, I connected ORM (clean install on RPi Bookworm) to my Android phone, and I get past this screen, receiving speed and strokerate. Please note, I follow the instructions to the letter, so a copy of OS lite 64Bit, and let the installer do the rest. I don't know if installing the desktop interferes here.

In the installer we do modify the Bluetooth stack (by 'sudo apt-get -y install bluetooth bluez libbluetooth-dev libudev-dev git') and people must do a
```
sudo setcap cap_net_bind_service,cap_net_raw=+eip $(eval readlink -f `which node`)
```

Could it be that something on the iPhone interferes here (Apple is known to not follow standards completely....)

**Comment by Abasz** [comment]
As I wrote in the other thred. One can use EFR to check what data is being broadcasted: https://github.com/laberning/openrowingmonitor/discussions/161#discussioncomment-8434225

Its avaialble for IOs too: https://apps.apple.com/us/app/efr-connect-ble-mobile-app/id1030932759
